src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -173,30 +183,107 @@ ReadFromMergeTree::ReadFromMergeTree(
     }
 }
 
-Pipe ReadFromMergeTree::readFromPool(
+
+Pipe ReadFromMergeTree::readFromPoolParallelReplicas(
     RangesInDataParts parts_with_range,
     Names required_columns,
     size_t max_streams,
     size_t min_marks_for_concurrent_read,
-    bool use_uncompressed_cache)
+    bool use_uncompressed_cache
+)
 {
+    const auto & client_info = context->getClientInfo();
+    auto extension = ParallelReadingExtension
+    {
+        .all_callback = all_ranges_callback.value(),
+        .callback = read_task_callback.value(),
+        .count_participating_replicas = client_info.count_participating_replicas,
+        .number_of_current_replica = client_info.number_of_current_replica,
+        .colums_to_read = required_columns
+    };
+
+    /// We have a special logic for local replica. It has to read less data, because in some cases it should
+    /// merge states of aggregate functions or do some other important stuff other than reading from Disk.
+    auto is_local_replica = context->getClientInfo().interface == ClientInfo::Interface::LOCAL;
+    if (!is_local_replica)
+        min_marks_for_concurrent_read = static_cast<size_t>(min_marks_for_concurrent_read * context->getSettingsRef().parallel_replicas_single_task_marks_count_multiplier);
+
+    auto pool = std::make_shared<MergeTreeReadPoolParallelReplicas>(
+        storage_snapshot,
+        max_streams,
+        extension,
+        parts_with_range,
+        prewhere_info,
+        required_columns,
+        virt_column_names,
+        min_marks_for_concurrent_read
+    );
+
     Pipes pipes;
-    size_t sum_marks = 0;
-    size_t total_rows = 0;
+    const auto & settings = context->getSettingsRef();
+    size_t total_rows = parts_with_range.getRowsCountAllParts();
 
-    for (const auto & part : parts_with_range)
+    for (size_t i = 0; i < max_streams; ++i)
     {
-        sum_marks += part.getMarksCount();
-        total_rows += part.getRowsCount();
+        auto algorithm = std::make_unique<MergeTreeThreadSelectAlgorithm>(
+            i, pool, min_marks_for_concurrent_read, max_block_size,
+            settings.preferred_block_size_bytes, settings.preferred_max_column_in_block_size_bytes,
+            data, storage_snapshot, use_uncompressed_cache,
+            prewhere_info, actions_settings, reader_settings, virt_column_names);
+
+        auto source = std::make_shared<MergeTreeSource>(std::move(algorithm));
+
+        /// Set the approximate number of rows for the first source only
+        /// In case of parallel processing on replicas do not set approximate rows at all.
+        /// Because the value will be identical on every replicas and will be accounted
+        /// multiple times (settings.max_parallel_replicas times more)
+        if (i == 0 && !client_info.collaborate_with_initiator)
```

Comment:
```suggestion
        if (i == 0 && !client_info.collaborate_with_initiator)
```
again do you need the check ` !client_info.collaborate_with_initiator`?
The method is called `readFromPoolParallelReplicas` so can you end up here if parallel replicas are not used?

Reply:
It means that we are reading with parallel replicas and we are on initiator node (local one). We need this to report total rows only one instead of N times.

Reply:
Got confused by the comment then (/// In case of **parallel processing on replicas** do not set approximate rows at all.)